hal/armv8m/pmap.c

```diff
@@ -105,11 +190,78 @@ int pmap_segment(unsigned int i, void **vaddr, size_t *size, int *prot, void **t
 
 void _pmap_init(pmap_t *pmap, void **vstart, void **vend)
 {
+	const syspage_map_t *ikmap;
+	unsigned int ikregion;
+	addr_t pc;
+	unsigned int i, cnt = syspage->hs.mpu.allocCnt;
+
 	(*vstart) = (void *)(((ptr_t)_init_vectors + 7) & ~7u);
 	(*vend) = (*((char **)vstart)) + SIZE_PAGE;
 
 	pmap->start = (void *)&__bss_start;
 
 	/* Initial size of kernel map */
 	pmap->end = (void *)((addr_t)&__bss_start + 32 * 1024);
+
+	/* Configure MPU */
+	pmap_common.mpu = MPU_BASE;
+
+	/* Disable MPU just in case */
+	*(pmap_common.mpu + mpu_ctrl) &= ~1;
+	hal_cpuDataMemoryBarrier();
+
+	/* Allow unlimited kernel access */
+	*(pmap_common.mpu + mpu_ctrl) |= (1 << 2);
+	hal_cpuDataMemoryBarrier();
+
+	for (i = 0; i < cnt; ++i) {
+		/* Select MPU region to configure */
+		*(pmap_common.mpu + mpu_rnr) = i;
+		hal_cpuDataMemoryBarrier();
+
+		*(pmap_common.mpu + mpu_rbar) = syspage->hs.mpu.table[i].rbar;
+		hal_cpuDataMemoryBarrier();
+
+		/* Disable regions for now */
+		*(pmap_common.mpu + mpu_rlar) = syspage->hs.mpu.table[i].rlar & ~1u;
+		hal_cpuDataMemoryBarrier();
+	}
+
+	/* Enable MemManage fault in SHCSR */
+	*((volatile unsigned int *)SHCSR_BASE) |= 1 << 16;
+
+	/* Enable MPU */
+	*(pmap_common.mpu + mpu_ctrl) |= 1;
+	hal_cpuDataMemoryBarrier();
+
+	/* FIXME HACK
+	 * allow all programs to execute (and read) kernel code map.
+	 * Needed because of hal_jmp, syscalls handler and signals handler.
+	 * In these functions we need to switch to the user mode when still
+	 * executing kernel code. This will cause memory management fault
+	 * if the application does not have access to the kernel instruction
+	 * map. Possible fix - place return to the user code in the separate
+	 * region and allow this region instead. */
+
+	/* Find kernel code region */
+	__asm__ volatile("\tmov %0, pc;" : "=r"(pc));
+	ikmap = syspage_mapAddrResolve(pc);
```

Comment:
I know it's copy-pasted from other code but I don't like this inline assembly - the same effect can be achieved with portable code.
```suggestion
	ikmap = syspage_mapAddrResolve((addr_t)_pmap_init);
```
`addr_t pc;` needs to be removed too.